README.md
# DQSEGDB2

[![PyPI version](https://badge.fury.io/py/dqsegdb2.svg)](http://badge.fury.io/py/dqsegdb2)
[![Conda version](https://img.shields.io/conda/vn/conda-forge/dqsegdb2.svg)](https://anaconda.org/conda-forge/dqsegdb2/)
[![License](https://img.shields.io/pypi/l/dqsegdb2.svg)](https://choosealicense.com/licenses/gpl-3.0/)
[![DOI](https://zenodo.org/badge/136390328.svg)](https://zenodo.org/badge/latestdoi/136390328)
![Supported Python versions](https://img.shields.io/pypi/pyversions/dqsegdb2.svg)

[![Build status](https://github.com/duncanmmacleod/dqsegdb2/actions/workflows/build.yml/badge.svg?branch=master)](https://github.com/duncanmmacleod/dqsegdb2/actions/workflows/build.yml)
[![Maintainability](https://api.codeclimate.com/v1/badges/da522356dd6a3d90eaac/maintainability)](https://codeclimate.com/github/duncanmmacleod/dqsegdb2/maintainability)
[![Documentation status](https://readthedocs.org/projects/dqsegdb2/badge/?version=latest)](https://dqsegdb2.readthedocs.io/en/latest/?badge=latest)
[![Coverage status](https://codecov.io/gh/duncanmmacleod/dqsegdb2/branch/master/graph/badge.svg)](https://codecov.io/gh/duncanmmacleod/dqsegdb2)

DQSEGDB2 is a simplified Python implementation of the DQSEGDB API as defined in
[LIGO-T1300625](https://dcc.ligo.org/LIGO-T1300625).

This package only provides a query interface for `GET` requests, any users
wishing to make `POST` requests should refer to the official `dqsegdb` Python
client available from <https://github.com/ligovirgo/dqsegdb/>.

## Installation

For best results, please install from [conda-forge](https://conda-forge.org/):

```shell
conda install -c conda-forge dqsegdb2
```

On Unix systems, you can also install directly from PyPI:

```shell
pip install dqsegdb2
```

## Basic Usage

```python
from dqsegdb2.query import query_segments
print(query_segments('G1:GEO-SCIENCE:1', 1000000000, 1000001000))
```


debian/changelog
dqsegdb2 (1.2.1-1) unstable; urgency=low

  * update for 1.2.1
  * update igwn-auth-utils minimum requirement

 -- Duncan Macleod <duncan.macleod@ligo.org>  Fri, 25 Aug 2023 08:54:00 +0100

dqsegdb2 (1.2.0-1) unstable; urgency=low

  * update for 1.2.0
  * update igwn-auth-utils minimum requirement

 -- Duncan Macleod <duncan.macleod@ligo.org>  Wed, 16 Aug 2023 10:48:00 +0100

dqsegdb2 (1.1.4-1) unstable; urgency=low

  * update for 1.1.4

 -- Duncan Macleod <duncan.macleod@ligo.org>  Tue, 23 May 2023 10:52:00 +0100

dqsegdb2 (1.1.3-1) unstable; urgency=low

  * update for 1.1.3
  * update igwn-auth-utils requirement
  * remove extra Requires for igwn-auth-utils[requests]

 -- Duncan Macleod <duncan.macleod@ligo.org>  Mon, 26 Sep 2022 15:53:00 +0100

dqsegdb2 (1.1.2-1) unstable; urgency=low

  * update for 1.1.2, reinstates debian packaging
  * remove python2 packages

 -- Duncan Macleod <duncan.macleod@ligo.org>  Thu, 30 Jun 2022 17:00:00 +0100

dqsegdb2 (1.0.1-1) unstable; urgency=low

  * first release

 -- Duncan Macleod <duncan.macleod@ligo.org>  Thu, 07 Feb 2019 13:57:00 +0000


dqsegdb2/query.py
# -*- coding: utf-8 -*-
# DQSEGDB2
# Copyright (C) 2018,2020,2022  Duncan Macleod
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""Query methods for DQSEGDB2.
"""

from functools import partial

from ligo import segments

from igwn_auth_utils.scitokens import target_audience as scitoken_audience

from . import api
from .requests import (
    Session,
    get_json,
)
from .utils import get_default_host


def _url(host, path_func, *args, **kwargs):
    """Construct the full URL for a query to ``host`` using an API path func.
    """
    path = path_func(*args, **kwargs)
    if host is None:
        host = get_default_host()
    return f"{host.rstrip('/')}/{path.lstrip('/')}"


def query_ifos(
    host=None,
    raw=False,
    **request_kwargs,
):
    """Query for all defined interferometers (IFOs).

    Parameters
    ----------
    host : `str`, optional
        The URL of the DQSegDB server; if `None`
        :func:`~dqsegdb2.utils.get_default_host` will be used to discover
        the default host.

    raw : `bool`, optional
        Return the full JSON response from the request.

    Returns
    -------
    If ``raw=False``:

    ifos : `set`
        the set of all known IFO prefices.

    If ``raw=True`` the full JSON response is returned.

    Examples
    --------
    >>> from dqsegdb2.query import query_ifos
    >>> query_ifos()
    {'H1', 'V1', 'G1', 'K1', 'L1'}
    >>> query_ifos(raw=True)
    {"query_information": {
         "start": 0,
         "server_timestamp": 1356794344,
         "end": 0,
         "server_elapsed_query_time": "0.00131",
         "include": [],
         "uri": "/dq",
         "api_version": "2.1.17",
         "server": "segments"},
     "Ifos": ["G1", "H1", "K1", "L1", "V1"]}
    """
    url = _url(host, api.ifos_path)
    out = get_json(url, **request_kwargs)
    if raw:
        return out
    return set(out["Ifos"])


def query_names(
    ifo,
    host=None,
    raw=False,
    **request_kwargs,
):
    """Query for all defined flags for the given ``ifo``

    Parameters
    ----------
    ifo : `str`
        The interferometer prefix for which to query.

    host : `str`, optional
        The URL of the DQSegDB server; if `None`
        :func:`~dqsegdb2.utils.get_default_host` will be used to discover
        the default host.

    raw : `bool`, optional
        Return the full JSON response from the request.

    Returns
    -------
    flags : `set`
        the set of all define flag names in the format ``{ifo}:{name}``

    Examples
    --------
    >>> from dqsegdb2.query import query_names
    >>> query_names('G1')
    """
    url = _url(host, api.flags_path, ifo)
    out = get_json(url, **request_kwargs)
    if raw:
        return out
    return {f'{ifo}:{name}' for name in out['results']}


def query_versions(flag, host=None, raw=False, **request_kwargs):
    """Query for defined versions for the given flag.

    Parameters
    ----------
    flag : `str`
        The name for which to query.

    host : `str`, optional
        The URL of the DQSegDB server; if `None`
        :func:`~dqsegdb2.utils.get_default_host` will be used to discover
        the default host.

    raw : `bool`, optional
        Return the full JSON response from the request.

    Returns
    -------
    versions : `list` of `int`
        the list of defined versions for the given flag

    Examples
    --------
    >>> from dqsegdb2.query import query_versions
    >>> query_versions('G1:GEO-SCIENCE')
    [1, 2, 3]
    """
    ifo, name = flag.split(':', 1)
    url = _url(host, api.versions_path, ifo, name)
    out = get_json(url, **request_kwargs)
    if raw:
        return out
    return sorted(out['version'])


def query_segments(
    flag,
    start,
    end,
    host=None,
    coalesce=True,
    raw=False,
    **request_kwargs,
):
    """Query for segments for the given flag in a ``[start, stop)`` interval.

    Parameters
    ----------
    flag : `str`
        The name for which to query, see _Notes_ for information on how
        versionless-flags are queried.

    start : `int`
        The GPS start time.

    end : `int`
        The GPS end time.

    host : `str`, optional
        The URL of the DQSegDB server; if `None`
        :func:`~dqsegdb2.utils.get_default_host` will be used to discover
        the default host.

    coalesce : `bool`, optional
        If `True`, coalesce the segmentlists returned by the server,
        and restrict them to lie fully within the ``[start, end)``
        request segment, otherwise return the 'raw' result,
        default: `True`.
        This option is ignored if ``raw=True`` is given.

    raw : `bool`, optional
        Return the full JSON response from the request.
        If an explicit version is not given, the result will be a
        `list` of JSON responses, one for each discovered version.

    Returns
    -------
    If ``raw=False`` is given:

    segmentdict : `dict`
        A `dict` with the following keys

        - ``'ifo'`` - the interferometer prefix (`str`)
        - ``'name'`` - the flag name (`str`)
        - ``'version'`` - the flag version (`int`)
        - ``'known'`` - the known segments (`~ligo.segments.segmentlist`)
        - ``'active'`` - the active segments (`~ligo.segments.segmentlist`)
        - ``'metadata'`` - a `dict` of flag information (`dict`)
        - ``'query_information'`` - a `dict` of query information (`dict`)

    If ``raw=True`` is given, **and** the flag name includes an explicit
    version, the result will be the raw JSON response from the single request,
    otherwise a `list` of JSON responses will be returned.

    Notes
    -----
    If ``flag`` is given without a version (e.g. ``'X1:FLAG-NAME'``) or the
    version is given as ``'*'`` (e.g. ``'X1:FLAG-NAME:*'``) the result of
    the query will be the intersection of queries over all versions found
    in the database.
    In that case the ``'metadata'`` and ``'query_information'`` in the output
    will be preserved for the highest version number only (if ``raw=False``).

    Examples
    --------
    >>> from dqsegdb2.query import query_segments
    >>> query_segments('G1:GEO-SCIENCE:1', 1000000000, 1000001000)
    """
    request = segments.segmentlist([
        segments.segment(float(start), float(end)),
    ])

    try:
        ifo, name, version = flag.split(':', 2)
        versions = [int(version)]
    except ValueError:
        single_version = False
        if flag.endswith(':*'):  # allow use of wildcard version
            flag = flag.rsplit(':', 1)[0]
        ifo, name = flag.split(':', 1)
    else:
        single_version = True

    # what to ask for:
    include = {"known", "active"}
    if raw:
        include.add("metadata")

    # construct partial func to use for each versioned URL
    _format_url = partial(
        _url,
        host,
        api.resources_path,
        ifo,
        name,
        s=start,
        e=end,
        include=",".join(sorted(include)),
    )

    # set default audience
    request_kwargs.setdefault(
        "token_audience",
        scitoken_audience(host or get_default_host()),
    )

    # use Session to query for and then loop over versions (if needed)
    with Session(**request_kwargs) as sess:

        if not single_version:  # query for all versions
            versions = sorted(query_versions(flag, host=host, session=sess))

        if raw and not single_version:
            out = []
        else:
            # custom subset of information, with at least the following
            # keys (and types)
            out = dict(
                known=segments.segmentlist(),
                active=segments.segmentlist(),
                ifo=ifo,
                name=name,
                version=versions[0],
            )

        for version in versions:
            url = _format_url(version)
            result = get_json(url, session=sess)

            if raw and single_version:
                return result

            if raw:
                out.append(result)
                continue

            # if not raw, convert to segment objects
            # and coalesce if asked
            for key in ('active', 'known'):
                out[key].extend(segments.segmentlist(map(
                    segments.segment,
                    result.pop(key),
                )))
                if coalesce:
                    out[key] = out[key].coalesce() & request
            out.update(result)

        if len(versions) > 1 and not raw:  # unset the version if multiple
            out["version"] = None

    return out
